test/bench/bench.cpp

```diff
@@ -0,0 +1,242 @@
+        [](auto x) noexcept { return !std::isspace(x); });
+    return result;
+}
+
+void benchmark_parse(benchmark::State& state, const fizzy::bytes& wasm_binary)
+{
+    const auto input_size = wasm_binary.size();
+    auto num_bytes_parsed = uint64_t{0};
+    for (auto _ : state)
+    {
+        const auto module = fizzy::parse(wasm_binary);
+        benchmark::DoNotOptimize(module);
+        num_bytes_parsed += input_size;
+    }
+    state.counters["size"] = benchmark::Counter(static_cast<double>(input_size));
+    state.counters["rate"] =
+        benchmark::Counter(static_cast<double>(num_bytes_parsed), benchmark::Counter::kIsRate);
+}
+
+void benchmark_instantiate(benchmark::State& state, const fizzy::bytes& wasm_binary)
+{
+    const auto module = fizzy::parse(wasm_binary);
+    for (auto _ : state)
+    {
+        auto instance = fizzy::instantiate(module, {});
+        benchmark::DoNotOptimize(instance);
+    }
+}
+
+struct ExecutionBenchmarkCase
+{
+    std::shared_ptr<const fizzy::bytes> wasm_binary;
+    std::string func_name;
+    std::vector<uint64_t> func_args;
+    fizzy::bytes memory;
+    std::optional<uint64_t> expected_result;
+    fizzy::bytes expected_memory;
+
+    void operator()(benchmark::State& state)
+    {
+        const auto module = fizzy::parse(*wasm_binary);
+        const auto func_idx = fizzy::find_exported_function(module, func_name);
+        if (!func_idx)
+            return state.SkipWithError(("Function \"" + func_name + "\" not found").c_str());
+
+        auto instance = fizzy::instantiate(module, {});
+        {
+            // TODO: Check if memory is exported or imported.
+            if (memory.size() > instance.memory.size())
+                return state.SkipWithError("Cannot init memory");
+            std::copy(std::begin(memory), std::end(memory), std::begin(instance.memory));
+            const auto result = fizzy::execute(instance, *func_idx, func_args);
+            if (result.trapped)
+                return state.SkipWithError("Trapped");
+
+            if (expected_result)
+            {
+                if (result.stack.size() != 1)
+                    return state.SkipWithError("Incorrect number of results");
+                else if (result.stack[0] != *expected_result)
+                    return state.SkipWithError("Incorrect result");
+            }
+            else if (!result.stack.empty())
+                return state.SkipWithError("Unexpected result");
+
+            if (instance.memory.size() < expected_memory.size())
+                return state.SkipWithError("Result memory is shorted than expected");
+
+            if (!std::equal(std::begin(expected_memory), std::end(expected_memory),
+                    std::begin(instance.memory)))
+                return state.SkipWithError("Incorrect result memory");
+        }
+
+        for (auto _ : state)
+        {
+            std::copy(std::begin(memory), std::end(memory), std::begin(instance.memory));
```

Comment:
This actually doesn't seem correct in case `execute` of previous iteration writes to the bytes above initial memory
(you'll reset only the first bytes, not the whole page)

Reply:
True. I will have to copy initial memory from the instance restore it for every loop iteration.